proposals/NNNN-inheritance-of-actor-isolation.md

```diff
@@ -0,0 +1,551 @@
+
+### Syntax sugar for inheriting actor isolation
+
+Isolated parameters have three downsides.
+
+The first downside is that the use pattern we expect to dominate ---
+declaring a function to inherit its caller's isolation --- is pretty
+cumbersome:
+
+```swift
+func foo(isolation: isolated (any Actor)? = #isolation)
+```
+
+The second downside is we can only do this if we can add formal
+parameters to a function.  Unfortunately, there are several places
+in the language where we really can't do that, most importantly
+accessors for computed properties:
+
+```swift
+var count: Int {
+  get { // How do we add an isolated parameter here?
+    ...
+  }
+}
+```
+
+The third downside is minor in comparison, but this pattern naturally
+turns into passing an actor reference, which isn't the most efficient
+way of passing down actor isolation because it still requires dynamic
+dispatch in order to extract the executor.  It would be better for the
+implementation if we could pass down the exact `UnownedSerialExecutor?`
+value that's needed at runtime.  While we do not currently want to
+encourage programmers to work with values of this type directly because
+of its tricky lifetime semantics, the compiler can manage it fairly
+easily.
+
+All of these downsides could be addressed by adding an attribute
+which causes an entity (including an accessor) to inherit its caller's
+isolation. This would be equivalent to receiving an `isolated` parameter
+with the same value as would be produced by `#isolated`, but it's easier
+to write, can be used in a few places that can't add arbitrary parameters, 
+and may be more efficiently implementable.
+
+### Isolated function types
+
+This proposal is focused on propagating isolation information *into*
+functions, but it's also interesting to look at propagating isolation
+*out* of functions.  Currently, the Swift type system only allows
+function isolation to be expressed in limited ways: functions can be
+declared as isolated to a global actor (e.g. `@MainActor () -> ()`), but
+all other kinds of isolation must be "type-erased", leaving a value
+whose type appears to be non-isolated.  This forces some of the same
+awkward `Sendable` restrictions that this proposal discusses with the
+`sequentialMap` example in the Motivation section.
+
+One way to solve this would be to introduce value-dependent isolated
+function types.  With such a feature, you could declare a value to have
+type, say, `@isolated(myActor) () -> ()`, where `myActor` is a `let`
+constant in the local scope.  This kind of value dependence, however,
+is a large step in complexity for a type system, and it's not a likely
+path for Swift in the foreseeable future.
+
+A more promising approach would be to allow the isolation to be
+statically erased but still make it dynamically recoverable by carrying
+it along in the function value, essentially as an extra value of type
+`(any Actor)?`.  This would look something like `@isolated () -> ()`,
```

Comment:
Drafting nit: "This" here, grammatically, refers to the preceding "value of type `(any Actor)?`," which clearly cannot be anything like `@isolated () -> ()`. Could the subject of this sentence be spelled out concretely?